pkg/storage/internal/actions/abort.go

```diff
@@ -0,0 +1,88 @@
+package actions
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/logging"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/storage/entity"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/storage/history"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wdk"
+)
+
+type abortAction struct {
+	logger           *slog.Logger
+	transactionsRepo TransactionsRepo
+}
+
+func newAbortAction(logger *slog.Logger, transactions TransactionsRepo) *abortAction {
+	return &abortAction{
+		logger:           logging.Child(logger, "abort_action"),
+		transactionsRepo: transactions,
+	}
+}
+
+func (a *abortAction) AbortAction(ctx context.Context, userID int, args *wdk.AbortActionArgs) (*wdk.AbortActionResult, error) {
+	txEntity, err := a.transactionsRepo.FindUniqueTransactionByReference(ctx, userID, *args.Reference)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find unique transaction by reference: %w", err)
+	}
+
+	if txEntity == nil && len(*args.Reference) == 64 {
+		txEntity, err = a.transactionsRepo.FindTransactionByUserIDAndTxID(ctx, userID, *args.Reference)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find transaction by txid: %w", err)
+		}
+	}
+
+	if txEntity == nil {
+		return nil, fmt.Errorf("failed to find unique transaction by reference: expected exactly one transaction with reference %s, found 0", *args.Reference)
+	}
+
+	err = validateTxForAbort(txEntity)
+	if err != nil {
+		return nil, err
+	}
+
+	history := history.NewBuilder().AbortAction(*args.Reference)
+
+	if txEntity.TxID != nil {
+		err = a.transactionsRepo.UpdateTransactionStatusForTxID(ctx, *txEntity.TxID, wdk.TxStatusFailed, wdk.ProvenTxStatusInvalid, history)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update known transaction status: %w", err)
+		}
+	}
+
+	err = a.transactionsRepo.UpdateTransactionStatusByID(ctx, txEntity.ID, wdk.TxStatusFailed, wdk.ProvenTxStatusInvalid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update transaction status: %w", err)
+	}
+
+	return &wdk.AbortActionResult{
+		Aborted: true,
+	}, nil
+}
+
+func validateTxForAbort(txEntity *entity.Transaction) error {
+	if !txEntity.IsOutgoing {
+		return fmt.Errorf("txDirectionInvalid: must be an outgoing transaction.")
+	}
+
+	return validateTxStatusForAbort(txEntity.Status)
+}
+
+func validateTxStatusForAbort(txStatus wdk.TxStatus) error {
+	var unAbortableStatuses = map[wdk.TxStatus]bool{
+		wdk.TxStatusCompleted: true,
+		wdk.TxStatusFailed:    true,
+		wdk.TxStatusSending:   true,
+		wdk.TxStatusUnproven:  true,
+	}
+
+	if unAbortableStatuses[txStatus] {
+		return fmt.Errorf("txNotAbortable: action with status %s cannot be aborted", txStatus)
```

Comment:
It looks like you had some intuition or something that you're trying to distinguish not abortable errors from others.

I would suggest to create a public error like:
```go
var ErrNotAbortableAction = fmt.Errorf("action cannot be aborted")
```


And then use it here:
```suggestion
		return fmt.Errorf("%w: action with status %s cannot be aborted", wdk.ErrNotAbortableAction,  txStatus)
```

Reply:
changed